core/src/main/java/org/elasticsearch/action/support/replication/ReplicationOperation.java

```diff
@@ -308,6 +308,11 @@ public static boolean isConflictException(Throwable t) {
         ShardRouting routingEntry();
 
         /**
+         * primary term for this primary (fixed when primary shard is created)
+         */
+        long primaryTerm();
```

Comment:
+1 to explore that option. Although, I always doubted whether we should set the primary terms on the ReplicationOperation level, I didn't do it because it wasn't needed and reluctantly opted to keep things as simple as possible and left it at that assert we have now. With the move the primary term based failures we bring the terms into scope, which makes me more inclined to also set them. As said, I'm still on the fence about it myself so I will go along with not doing it, if you feel differently.